### ISSUE TITLE
Replace button doesn’t work

### ISSUE THREAD

**Issue by CameronSamuels** [thecoderators/memerator#54]
The replace doesn’t work.

**Comment by CameronSamuels** [comment]
This might be able to be fixed by using `textContent` for the upload button. Either way, using this will improve the speed and make the app be more like what a website should be like (not using "unsafe" techniques like changing the `innerHTML`)

**Comment by oleteacher** [comment]
Not sure I follow "might be able to be fixed by using textContent for the upload button."?

Could you give me a tip so could try the fix?

Really cool meme generator and love to use:)

**Comment by CameronSamuels** [comment]
@oleteacher, Using `element.textContent` is more recommended than `element.innerHTML` because using `innerHTML` makes the browser have to re-render the HTML code where `textContent` does not.

So we should do this even if it doesn’t fix the problem.

The problem is that it is replacing the html (line 24 if script.js) by using `$('#upload').html()`.

I believe this is causing EventListener is removed because of the re-rendering (from replacing the html with the same code but with “Replace” but making it think it’s different nodes).

Using `textContent` can fix this because it keeps the nodes and doesn’t re-render, it just changes the text.

You will need to use `$('#upload').text()` to set the initial text to “Upload” on page load instead of in the html file.

Thanks for your help!

**Comment by oleteacher** [comment]
Thank you for reply @CameronSamuels 

Not getting far in my testing but did discover can replace image by commenting out part of line 24:

`$('.row').css('visibility', 'visible'); //$('#upload').html('Replace' + $('#upload').html().replace('Upload', ''));`

Image is replaced. You have to restart the text and error is produced in console:

```
Uncaught TypeError: Cannot set property 'id' of null
    at Image.img.onload (script.js?v2.1:14)
```
Will play around some more tomorrow during play time:)

**Comment by CameronSamuels** [comment]
@oleteacher, removing that line does solve the problem, but the text still says "Upload", and it would make sense to have it say "Replace".

The reason for the error you had is because of line 14 setting `#download, #advanced` to have a `""` `id`. So the second time it is uploaded, it can't find `#download` or `#advanced` because those elements don't have that `id` anymore.

If the `EventListener` was working, then that error would have shown, but it didn't show because the function never executed due to it not having an event anymore (like mentioned in the previous comment).

So to solve this problem, we would need to add a try statement to those statements of code (line 14). If we don't use a try statement, we can use a variable to determine if the code has already been executed. This could be most efficient by using the `textContent` property of `#upload` and seeing if it `== "Upload"` instead of `"Replace"`.

Using `textContent` like I was saying in the previous content will allow us to fix those errors, as well as fix the replacing button that does not work. It will also improve the browsers performance and usage of recommended strategies. We will definitely use `textContent` to resolve this issue as it should be done even if not to resolve the issue (which doing one does the other).

Thanks for your help! I will be glad to see you solve this issue.

**Comment by oleteacher** [comment]
Guess I am not the person to help with this. Have tried many things without any real fix. Sorry...

Not giving up but maybe someone else with more talent can solve this issue:)

**Comment by CameronSamuels** [comment]
It's okay. We aren't asking for you to do this, so no worries.

**Comment by oleteacher** [comment]
Yes of course I understand you are not asking me to do this. Just did not want you to think had forgotten, that is all.

Cheers!

**Comment by Pipe-to-Grep** [comment]
@CameronSamuels What if you tie a function that clears the canvas to the replace button? It's worth a shot

**Comment by CameronSamuels** [comment]
@Pipe-to-Grep, that won’t work. The “replace” button click doesn’t do anything. You can see what I mean by adding an `alert()` to the upload function and it will work for upload but will not work for replace. Because it doesn’t activate any code. It’s what I said in the previous comments.
